File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n=nums.length;
        int r=0;
        int ar[]= new int[n-k+1];
         
        Deque<Integer> dq= new ArrayDeque<>();

        for(int i=0;i<n;i++){

           if(!dq.isEmpty()&&dq.peek()==i-k)dq.poll();

            while(!dq.isEmpty()&&nums[dq.peekLast()]<nums[i])
            dq.pollLast();
           
            dq.offer(i);
            if(i>=k-1)
            ar[r++]=nums[dq.peek()];


        }

       
        return ar;
    }
}

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        deque<int> dq;
        int n = nums.size();

        for(int i=0;i<n;i++){
            if(!dq.empty() && dq.front() <= i-k)dq.pop_front(); // ye dekhta hai ki out of index wale element na rhe

            while(!dq.empty() && nums[dq.back()] <= nums[i])dq.pop_back();// ye monotonic deque(or priority queue) banne ka try kr rhe
            dq.push_back(i);//pushing index

            if(i >= k-1)ans.push_back(nums[dq.front()]);// store result AS front me ans hai
        }
        return ans;
    
    }
};